lib/active_merchant/billing/gateways/forte.rb

```diff
@@ -0,0 +1,246 @@
+          post[:billing_address][:address_zip] = address[:zip] if address[:zip]
+          post[:billing_address][:address_state] = address[:state] if address[:state]
+          post[:billing_address][:address_city] = address[:city] if address[:city]
+        end
+      end
+
+      def add_shipping_address(post, options)
+        return unless options[:shipping_address]
+        address = options[:shipping_address]
+
+        post[:shipping_address] = {}
+        post[:shipping_address][:first_name] = address[:name].split(" ").first if address[:name]
+        post[:shipping_address][:last_name] = address[:name].split(" ").last if address[:name]
+        post[:shipping_address][:address_line1] = address[:address1] if address[:address1]
+        post[:shipping_address][:address_line2] = address[:address2] if address[:address2]
+        post[:shipping_address][:address_country] = address[:country] if address[:country]
+        post[:shipping_address][:address_zip] = address[:zip] if address[:zip]
+        post[:shipping_address][:address_state] = address[:state] if address[:state]
+        post[:shipping_address][:address_city] = address[:city] if address[:city]
+      end
+
+      def add_payment_method(post, payment_method)
+        if payment_method.respond_to?(:brand)
+          add_credit_card(post, payment_method)
+        else
+          add_echeck(post, payment_method)
+        end
+      end
+
+      def add_echeck(post, payment)
+        post[:echeck] = {}
+        post[:echeck][:account_holder] = payment.name
+        post[:echeck][:account_number] = payment.account_number
+        post[:echeck][:routing_number] = payment.routing_number
+        post[:echeck][:account_type] = payment.account_type
+        post[:echeck][:check_number] = payment.number
+      end
+
+      def add_credit_card(post, payment)
+        post[:card] = {}
+        post[:card][:card_type] = format_card_brand(payment.brand)
+        post[:card][:name_on_card] = payment.name
+        post[:card][:account_number] = payment.number
+        post[:card][:expire_month] = payment.month
+        post[:card][:expire_year] = payment.year
+        post[:card][:card_verification_value] = payment.verification_value
+      end
+
+      def commit(type, parameters)
+        add_auth(parameters)
+
+        url = (test? ? test_url : live_url)
+        response = parse(handle_resp(raw_ssl_request(type, url + endpoint, parameters.to_json, headers)))
+
+        Response.new(
+          success_from(response),
+          message_from(response),
+          response,
+          authorization: authorization_from(response),
+          avs_result: AVSResult.new(code: response[:response][:avs_result]),
+          cvv_result: CVVResult.new(response[:response][:cvv_code]),
+          test: test?,
+          error_code: error_code_from(response)
+        )
+      end
+
+      def handle_resp(response)
+        case response.code.to_i
+        when 200..400
```

Comment:
I would think you'd want `200..499` to get things like their funky 401 `This user does not have permission to access one or more of the requested Ids/Tokens.`